### PR TITLE
lufia2ac: coop support

### PR DESCRIPTION
**Pull request by el-u**
## What is this fixing or adding?

* **lufia2ac: coop support**
  This adds multiworld coop support to the Lufia II Ancient Cave world.
  The coop aspect applies both to opening blue chests to send out checks as well as to the collection of items for the Iris treasure hunt goals.
  - Blue chests opened by anyone connected to the coop slot will be added together, e.g., if personA has opened 3 blue chests while personB has opened 4 blue chests, then a total of 7 location checks will have been sent out from this slot.
  - Regarding the local Iris treasures, if, e.g., personA finds the Iris sword and Iris ring while personB finds the Iris armor, then all 3 of these Iris items will be counted towards the shared goal.

* **Core: update version to 0.4.2**
  This is required because the new feature can only be realized using a new version of the game client

* **CommonClient: add a framework for clients to subscribe to data storage key notifications**
Adds a new method `set_notify` to CommonContext. Using this method, a game client can subscribe to be notified of changes to selected data storage keys. The values of interest can be accessed via the new `stored_data` attribute of the context. It is a dictionary mapping the names of the data storage keys to their values. This dict is automatically kept up to date whenever the server sends out a packet containing the value of one of the subscribed keys.

  This framework is used to implement the new feature in the client of Lufia II. In principle it should be possible to rewrite FactorioClients usage of the "EnergyLink" key to use this new generic mechanism, but I did not attempt this since I don't play Factorio and wouldn't be able to test it.

* **Core: typing for async_start**
This PR also contains a correction to the type hints of the `async_start` method (which is used by the framework described above). The existing type hints only supported `bool` return values, even though there are already various existing uses with other return types (mostly `None`). Also, it permitted arbitrary yield and send types, even though these features don't seem to be usable through `async_start`.

## How was this tested?

https://discord.com/channels/731205301247803413/1112481808160923761
https://discord.com/channels/731205301247803413/1116393188702306424

### CODE REVIEW

File: Utils.py [hunk 1]
```diff
@@ -766,10 +766,10 @@ def read_snes_rom(stream: BinaryIO, strip_header: bool = True) -> bytearray:
     return buffer
 
 
-_faf_tasks: "Set[asyncio.Task[None]]" = set()
+_faf_tasks: "Set[asyncio.Task[typing.Any]]" = set()
 
 
-def async_start(co: Coroutine[typing.Any, typing.Any, bool], name: Optional[str] = None) -> None:
+def async_start(co: Coroutine[None, None, typing.Any], name: Optional[str] = None) -> None:
```
Comment (beauxq) [inline Utils.py:772]: This looks like good typing around this function.

**Review by black-sliver — COMMENTED** [review]
I checked everything that is not CommonClient.py or Client.py. Hoping @Berserker66 will find the time to do that

**Review by Berserker66 — APPROVED** [review]
Looked at Client.py and CommonClient.py
Wondering if the := syntax will result in people asking what it is on discord, as it's so rarely used, might even be the first time in AP.

File: Utils.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def as_simple_string(self) -> str:
         return ".".join(str(item) for item in self)
 
 
-__version__ = "0.4.1"
+__version__ = "0.4.2"
```
Comment (ThePhar) [inline Utils.py:45]: guess if there's any time to increment this, may as well do it now lol

**Review by ThePhar — APPROVED** [review]

**Pull request merged**